pinot-server/src/main/java/org/apache/pinot/server/starter/helix/OffsetBasedConsumptionStatusChecker.java

```diff
@@ -0,0 +1,119 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class is used at startup time to have a more accurate estimate of the catchup period in which no query execution
+ * happens and consumers try to catch up to the latest messages available in streams.
+ * To achieve this, every time status check is called, {@link #haveAllConsumingSegmentsReachedStreamLatestOffset},
+ * for each consuming segment, we check if segment's latest ingested offset has reached the latest stream offset.
+ * To prevent chasing a moving target, once the latest stream offset is fetched, it will not be fetched again and
+ * subsequent status check calls compare latest ingested offset with the already fetched stream offset.
+ */
+public class OffsetBasedConsumptionStatusChecker {
+  private static final Logger LOGGER = LoggerFactory.getLogger(OffsetBasedConsumptionStatusChecker.class);
+
+  // constructor parameters
+  private final InstanceDataManager _instanceDataManager;
+  private final Set<String> _consumingSegments;
+
+  // helper variable
+  private final Set<String> _caughtUpSegments = new HashSet<>();
+
+  public OffsetBasedConsumptionStatusChecker(InstanceDataManager instanceDataManager, Set<String> consumingSegments) {
+    _instanceDataManager = instanceDataManager;
+    _consumingSegments = consumingSegments;
+  }
+
+  public boolean haveAllConsumingSegmentsReachedStreamLatestOffset() {
+    boolean allSegsReachedLatest = true;
+    for (String segName : _consumingSegments) {
+      if (_caughtUpSegments.contains(segName)) {
+        continue;
+      }
+      TableDataManager tableDataManager = getTableDataManager(segName);
+      if (tableDataManager == null) {
+        LOGGER.info("TableDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        return false;
+      }
+      SegmentDataManager segmentDataManager = tableDataManager.acquireSegment(segName);
+      if (segmentDataManager == null) {
+        LOGGER.info("SegmentDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        return false;
+      }
+      if (!(segmentDataManager instanceof LLRealtimeSegmentDataManager)) {
+        // There's a possibility that a consuming segment has converted to a committed segment. If that's the case,
+        // segment data manager will not be of type LLRealtime.
+        LOGGER.info("Segment {} is already committed and is considered caught up.", segName);
+        _caughtUpSegments.add(segName);
+        releaseSegment(tableDataManager, segmentDataManager);
+        continue;
+      }
+      LLRealtimeSegmentDataManager rtSegmentDataManager = (LLRealtimeSegmentDataManager) segmentDataManager;
+      StreamPartitionMsgOffset latestIngestedOffset = rtSegmentDataManager.getCurrentOffset();
+      StreamPartitionMsgOffset latestStreamOffset = rtSegmentDataManager.getLatestStreamOffsetAtStartupTime();
+      releaseSegment(tableDataManager, segmentDataManager);
```

Comment:
Isnt it better to do release in a try/finally, so that in case there are some exceptions, we still release the segment?

Reply:
Done.